src/core/slice.c

```diff
@@ -347,6 +348,55 @@ static void slice_enumerate_perpetual(Manager *m) {
                 (void) slice_make_perpetual(m, SPECIAL_SYSTEM_SLICE, NULL);
 }
 
+static int slice_freezer_action(Unit *s, FreezerAction action) {
+        int r;
+        _cleanup_closedir_ DIR *d = NULL;
+
+        assert(s);
+        assert_return(IN_SET(action, FREEZER_FREEZE, FREEZER_THAW), -EINVAL);
+
+        r = cg_enumerate_subgroups(SYSTEMD_CGROUP_CONTROLLER, s->cgroup_path, &d);
+        if (r < 0)
+                return log_unit_error_errno(s, r, "Failed to open cgroup directory: %m");
+
+        for (;;) {
+                _cleanup_free_ char *subgroup = NULL;
+                Unit *u;
+
+                r = cg_read_subgroup(d, &subgroup);
+                if (r < 0)
+                        return log_unit_error_errno(s, r, "Failed to read list of sub-cgroups: %m");
+                if (r == 0)
+                        break;
+
+                u = manager_get_unit(s->manager, subgroup);
+                assert(u);
+
+                if (action == FREEZER_FREEZE && UNIT_VTABLE(u)->freeze)
+                        UNIT_VTABLE(u)->freeze(u);
+                else if (action == FREEZER_THAW && UNIT_VTABLE(u)->thaw)
+                        UNIT_VTABLE(u)->thaw(u);
```

Comment:
This does nothing for the children that do not support the operation. I'm not sure if that is OK. We would refuse freeze/thaw operations for unsupported unit types, but allow them on the parent slice, and do nothing. Wouldn't it be more appropriate to first loop over all children, check if they support the method, and refuse the whole operation if it cannot be done?

Reply:
OK, I will recursively check the slice before starting the operation on child units.